x/qgb/keeper/msg_server.go

```diff
@@ -45,8 +47,53 @@ func (k msgServer) ValsetConfirm(c context.Context, msg *types.MsgValsetConfirm)
 }
 
 // DataCommitmentConfirm handles MsgDataCommitmentConfirm
-func (k msgServer) DataCommitmentConfirm(context.Context, *types.MsgDataCommitmentConfirm) (*types.MsgDataCommitmentConfirmResponse, error) {
-	// TODO
+func (k msgServer) DataCommitmentConfirm(c context.Context, msg *types.MsgDataCommitmentConfirm) (*types.MsgDataCommitmentConfirmResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	sigBytes, err := hex.DecodeString(msg.Signature)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrInvalid, "signature decoding")
+	}
+
+	// verify validator address
+	validatorAddress, err := sdk.AccAddressFromBech32(msg.ValidatorAddress)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrInvalid, "validator address invalid")
+	}
+	validator, found := k.GetOrchestratorValidator(ctx, validatorAddress)
+	if !found {
+		return nil, sdkerrors.Wrap(types.ErrUnknown, "validator")
+	}
+	if err := sdk.VerifyAddressFormat(validator.GetOperator()); err != nil {
+		return nil, sdkerrors.Wrapf(err, "discovered invalid validator address for validator %v", validatorAddress)
+	}
+
+	// verify ethereum address
+	ethAddress, err := types.NewEthAddress(msg.EthAddress)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrInvalid, "invalid eth address")
+	}
+	err = types.ValidateEthereumSignature([]byte(msg.Commitment), sigBytes, *ethAddress)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrInvalid, fmt.Sprintf("signature verification failed expected sig by %s with checkpoint %s found %s", ethAddress, msg.Commitment, msg.Signature))
+	}
+	ethAddressFromStore, found := k.GetEthAddressByValidator(ctx, validator.GetOperator()) // check what to do with operator
+	if !found {
+		return nil, sdkerrors.Wrap(types.ErrEmpty, "no eth address set for validator")
+	}
+	if *ethAddressFromStore != *ethAddress {
+		return nil, sdkerrors.Wrap(types.ErrInvalid, "submitted eth address does not match delegate eth address")
+	}
```

Comment:
nice, these are important to check! but I think they would better fit in the `ValidateBasic` method for `MsgDataCommitmentConfirm`. In general, if we can perform a quick stateless check on a message, it should be in `ValidateBasic`

Reply:
We're doing the stateless ones in `ValidateBasic`. However, these require access to the state:
- `validator,` found := k.GetOrchestratorValidator(ctx, validatorAddress)`
- `ethAddressFromStore, found := k.GetEthAddressByValidator(ctx, validator.GetOperator())`

Reply:
makes sense